lib/label.ml

```diff
@@ -0,0 +1,191 @@
+        let op = of_string (Re.Group.get g 2) in
+        let value = Re.Group.get g 3 in
+        (label, Some (op, value))
+      with Not_found -> (s, None)
+end
+
+type t =
+  | Dir of string
+  | Source_tree of string
+  | File of string
+  | Part of string
+  | Env of string
+  | Skip
+  | Non_det of [`Output | `Command]
+  | Version of Relation.t * Ocaml_version.t
+  | Require_package of string
+  | Set of string * string
+  | Unset of string
+
+let pp ppf = function
+  | Dir d -> Fmt.pf ppf "dir=%s" d
+  | Source_tree s -> Fmt.pf ppf "source-tree=%s" s
+  | File f -> Fmt.pf ppf "file=%s" f
+  | Part p -> Fmt.pf ppf "part=%s" p
+  | Env e -> Fmt.pf ppf "env=%s" e
+  | Skip -> Fmt.string ppf "skip"
+  | Non_det `Output -> Fmt.string ppf "non-deterministic=output"
+  | Non_det `Command -> Fmt.string ppf "non-deterministic=command"
+  | Version (op, v) ->
+    Fmt.pf ppf "version%a%a" Relation.pp op Ocaml_version.pp v
+  | Require_package p -> Fmt.pf ppf "require-package=%s" p
+  | Set (v, x) -> Fmt.pf ppf "set-%s=%s" v x
+  | Unset x -> Fmt.pf ppf "unset-%s" x
+
+let is_prefix ~prefix s =
+  let len_prefix = String.length prefix in
+  if String.length s > len_prefix then
+    String.equal (String.sub s 0 len_prefix) prefix
+  else false
+
+(* [is_prefix ~prefix s] is always checked before. *)
+let split_prefix ~prefix s =
+  let len_prefix = String.length prefix in
+  String.sub s len_prefix (String.length s - len_prefix)
+
+let doesnt_accept_value ~label ~value res =
+  match value with
+  | Some _ ->
+    Error (`Msg (Format.sprintf "Label `%s` does not allow a value." label))
+  | None -> Ok res
+
+let requires_value ~label ~value f =
+  match value with
+  | Some (op, v) -> f op v
+  | None -> Error (`Msg (Format.sprintf "Label `%s` requires a value." label))
+
+let requires_eq ~label ~op ~value f =
+  match op with
+  | Relation.Eq -> f value
+  | _ ->
+    let msg =
+      Format.sprintf
+        "Label `%s` requires assignment using the `=` operator." label
+    in
+    Error (`Msg msg)
+
+let requires_eq_value ~label ~value f =
+  requires_value ~label ~value (fun op value ->
+      requires_eq ~label ~op ~value (fun x -> Ok (f x)))
+
+let interpret label value =
+  match label with
+  (* flags: labels without value *)
+  | "skip" ->
+    doesnt_accept_value ~label ~value Skip
+  | v when is_prefix ~prefix:"unset-" v ->
+    doesnt_accept_value ~label ~value (Unset (split_prefix ~prefix:"unset-" v))
+  (* labels requiring a value and any comparison operator *)
+  | "version" -> (
+      requires_value ~label ~value (fun op v ->
+          match Ocaml_version.of_string v with
+          | Ok v -> Ok (Version (op, v))
+          | Error (`Msg e) ->
+            let msg = Format.sprintf "Invalid `version` label value: %s." e in
+            Error (`Msg msg) ) )
+  (* non-deterministic accepts any value *)
+  | "non-deterministic" -> (
+      match value with
+      | Some (op, value) ->
+        requires_eq ~label ~op ~value (function
```

Comment:
I'm reverting e4ad2c2

Reply:
I think that is a mistake.
- The logic for the error "requires assignment" is duplicated
- The code interpreting that label is not on the same level of abstractions as the others
  (`requires_* ... (fun x -> interpret only the value)`, you have to care about relations and errors)
- `requires_eq` is needed to implement `requires_eq_value` (and it's still here)

Reply:
Non-det is a unique case here so I think it's okay if it differs.

I think the op logic is just `match op with Eq -> Ok .. | _ -> Error ...` so it's ok to duplicate it if it makes the code clearer.

`requires_eq` is not required by `requires_eq_value` it's just an indirection used to share the code between `requires_eq_value` and the non-det case. It can be inlined in `requires_eq_value`.

Reply:
The error formatting code could be shared, we could extract a `label -> value -> (_, string) result` for the wrong value error and a `label -> (_, string) result` for the non-eq op error.

That way we ensure consistent error messages while still clarifying the control flow.

Reply:
cc @gpetiot, what do you think?

Reply:
I thing we should not over-engineer this function (I think it's getting out of hand :D), only to not repeat an error message, I'd better keep it simple